File: scripts/lib/project.mjs
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import matter from 'gray-matter';
import { getSectionSlugSet } from './sidebar.mjs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export const ROOT_DIR = path.resolve(__dirname, '..', '..');
export const PROJECT_ROOT = ROOT_DIR;
export const DOCS_DIR = path.join(ROOT_DIR, 'src', 'content', 'docs');
export const SIDEBAR_PATH = path.join(ROOT_DIR, 'docs', 'SIDEBAR_URLS.md');

function normalizeMatchValue(value) {
  return String(value ?? '')
    .normalize('NFKC')
    .replace(/\s+/g, ' ')
    .trim()
    .toLowerCase();
}

export function findMdFiles(dir = DOCS_DIR) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  const files = [];

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files.push(...findMdFiles(fullPath));
      continue;
    }
    if (entry.name.endsWith('.md')) {
      files.push(fullPath);
    }
  }

  return files;
}

export function toRelativeDocPath(filePath) {
  return path.relative(ROOT_DIR, filePath);
}

export function getDocSection(relativePath) {
  const parts = relativePath.split(path.sep);
  return parts[3] ?? '';
}

// サイドバー解決のキャッシュ（成功時のみ保存。失敗はキャッシュせずリトライ可能）
const _sectionCache = new Map();
const DOCS_PREFIX = path.join('src', 'content', 'docs') + path.sep;

export function matchesSectionFilter(relativePath, data, sectionFilter) {
  if (!sectionFilter) return true;

  // サイドバーバックド解決（エイリアス対応・成功キャッシュ付き）
  let slugSet = _sectionCache.get(sectionFilter);
  if (!slugSet) {
    try {
      slugSet = getSectionSlugSet(sectionFilter);
      _sectionCache.set(sectionFilter, slugSet);
    } catch (e) {
      if (e.message?.startsWith('Unknown section')) {
        console.warn(`matchesSectionFilter: ${e.message} — falling back to heuristic match`);
      } else {
        console.warn(`matchesSectionFilter: unexpected error: ${e.message}`);
      }
    }
  }

  if (slugSet) {
    const rel = relativePath.startsWith(DOCS_PREFIX)
      ? relativePath.slice(DOCS_PREFIX.length).replace(/\.md$/, '')
      : path.basename(relativePath, '.md');
    return slugSet.has(rel);
  }

  // サイドバーに未登録のセクション名 — ヒューリスティックフォールバック
  const target = normalizeMatchValue(sectionFilter);
  if (!target) return true;

  const candidates = [
    relativePath,
    getDocSection(relativePath),
    data?.category,
    path.basename(relativePath, '.md'),
  ]
    .filter(Boolean)
    .map(normalizeMatchValue);

  return candidates.some((candidate) => candidate.includes(target));
}

/**
 * Convert an absolute .md file path to its path-based slug.
 * e.g. "/…/src/content/docs/overview/testim-overview.md" → "overview/testim-overview"
 */
export function filePathToSlug(filePath, docsDir = DOCS_DIR) {
  return path.relative(docsDir, filePath).replace(/\.md$/, '');
}

/**
 * Lazy-cached basename → path-slug lookup built from the docs index.
 * Values are `null` for ambiguous basenames (those appearing in multiple folders).
 * Cache is keyed by docsDir to support test isolation.
 */
const _basenameMapCache = new Map();
export function buildBasenameToPathMap(docsDir = DOCS_DIR) {
  const cached = _basenameMapCache.get(docsDir);
  if (cached) return cached;
  const slugIndex = buildSlugIndex(docsDir);
  const map = new Map();
  for (const slug of Object.keys(slugIndex)) {
    const bn = slug.split('/').pop();
    if (map.has(bn)) {
      map.set(bn, null); // ambiguous — skip
    } else {
      map.set(bn, slug);
    }
  }
  _basenameMapCache.set(docsDir, map);
  return map;
}

/**
 * Lazy-cached full slug → resolved full slug lookup.
 *
 * `resolveToFullSlug` is used from the parity extraction hot path, so repeated
 * lookups for the same truncated slug are memoized per docsDir. The cache must
 * stay under `resetProjectCachesForTest()` so test-only callers can force a
 * fresh scan after mutating a temp docs tree.
 */
const _resolveToFullSlugCache = new Map();

/**
 * Reset all module-level caches. Test-only API.
 */
export function resetProjectCachesForTest() {
  _sectionCache.clear();
  _basenameMapCache.clear();
  _slugIndexCache.clear();
  _resolveToFullSlugCache.clear();
}

/**
 * Lazy-cached full slug → { categoryFolder, filePath } index.
 *
 * Issue #247 re-review 第三弾 — `resolveToFullSlug` が `extractInvariantTokens`
 * → `createSegment` の hot path で呼ばれるため、毎回 repo 全体を再帰走査する
 * と full parity が main 比で倍以上遅くなっていた (測定で 4.32s → 9.79s)。
 * `buildBasenameToPathMap` と同じ docsDir-keyed Map で memoize して再帰走査を
 * 1 回に抑える。test 分離は `resetProjectCachesForTest` でクリアする。
 */
const _slugIndexCache = new Map();
export function buildSlugIndex(docsDir = DOCS_DIR) {
  const cached = _slugIndexCache.get(docsDir);
  if (cached) return cached;
  /** @type {Record<string, {categoryFolder:string, filePath:string}>} */
  const index = {};
  const walk = (dir) => {
    for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, ent.name);
      if (ent.isDirectory()) walk(full);
      else if (ent.isFile() && ent.name.endsWith('.md')) {
        const slug = filePathToSlug(full, docsDir);
        const categoryFolder = path.basename(path.dirname(full));
        index[slug] = { categoryFolder, filePath: full };
      }
    }
  };
  walk(docsDir);
  _slugIndexCache.set(docsDir, index);
  return index;
}

/**
 * Resolve a possibly truncated slug to the full slug present in the docs tree.
 *
 * Resolution order:
 *   1. Exact full-slug match in the docs index
 *   2. Unique basename fallback
 *   3. Original slug (safe fallback for ambiguous / missing entries)
 *
 * Cache is keyed by docsDir so temp test trees stay isolated from the real
 * repository docs tree.
 *
 * @param {string} slug
 * @param {string} [docsDir]
 * @returns {string}
 */
export function resolveToFullSlug(slug, docsDir = DOCS_DIR) {
  let dirCache = _resolveToFullSlugCache.get(docsDir);
  if (!dirCache) {
    dirCache = new Map();
    _resolveToFullSlugCache.set(docsDir, dirCache);
  }

  const cached = dirCache.get(slug);
  if (cached !== undefined) return cached;

  const index = buildSlugIndex(docsDir);
  let resolved;
  if (Object.prototype.hasOwnProperty.call(index, slug)) {
    resolved = slug;
  } else {
    const basename = slug.split('/').pop();
    const basenameMap = buildBasenameToPathMap(docsDir);
    resolved = basenameMap.get(basename) ?? slug;
  }

  dirCache.set(slug, resolved);
  return resolved;
}

/**
 * Resolve a CLI --slug value to a path-based slug.
 * Accepts both basename ("testim-overview") and path-based ("overview/testim-overview").
 * Returns null if the slug is not found or is ambiguous (logs a warning for ambiguity).
 *
 * **Deprecated**: Basename resolution is deprecated. Use path-based slugs directly
 * (e.g., `--slug=overview/testim-overview` instead of `--slug=testim-overview`).
 *
 * @param {string | null | undefined} input
 * @param {string} [docsDir]
 * @returns {string | null}
 */
export function resolveSlug(input, docsDir = DOCS_DIR) {
  if (!input) return null;
  const index = buildSlugIndex(docsDir);
  // Exact match (already path-based)
  if (index[input]) return input;
  // Basename resolution (deprecated): find all entries whose basename matches
  const basename = input.includes('/') ? null : input;
  if (!basename) return null;
  const matches = Object.keys(index).filter((slug) => slug.split('/').pop() === basename);
  if (matches.length === 1) {
    console.warn(
      `⚠️  Deprecated: basename slug "${input}" resolved to "${matches[0]}". Use the full path-based slug instead.`
    );
    return matches[0];
  }
  if (matches.length > 1) {
    console.warn(
      `⚠️  Ambiguous slug "${input}" matches multiple paths: ${matches.join(', ')}. Use full path.`
    );
    return null;
  }
  return null;
}

const SOURCE_URL_RE =
  /^https:\/\/docs\.tricentis\.com\/testim\/content\/([a-z0-9_-]+(?:\/[a-z0-9_-]+)*)\.htm$/;

/**
 * Extract the EN content path from a sourceUrl.
 *
 * Examples:
 *   ".../content/running-tests/play-from-here.htm"          → "running-tests/play-from-here"
 *   ".../content/running-tests/play-from-here/index.htm"    → "running-tests/play-from-here"
 *   ".../content/overview/testim-overview/use-ai/index.htm"  → "overview/testim-overview/use-ai"
 *
 * Returns null for non-matching URLs or non-string input.
 * @param {string | undefined | null} sourceUrl
 * @returns {string | null}
 */
export function extractSourceContentPath(sourceUrl) {
  if (typeof sourceUrl !== 'string') return null;
  const m = SOURCE_URL_RE.exec(sourceUrl);
  if (!m) return null;
  const raw = m[1];
  return raw.endsWith('/index') ? raw.slice(0, -'/index'.length) : raw;
}

/**
 * Richer doc index that includes the EN source content path from frontmatter.
 * Keys are path-based slugs (e.g., "overview/testim-overview").
 * @param {string} [docsDir]
 * @returns {Record<string, {filePath:string, localFolder:string, sourceContentPath:string|null}>}
 */
export function buildDocsIndex(docsDir = DOCS_DIR) {
  /** @type {Record<string, {filePath:string, localFolder:string, sourceContentPath:string|null}>} */
  const index = {};
  const walk = (dir) => {
    for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
      const full = path.join(dir, ent.name);
      if (ent.isDirectory()) {
        walk(full);
        continue;
      }
      if (!ent.isFile() || !ent.name.endsWith('.md')) continue;
      const slug = filePathToSlug(full, docsDir);
      const localFolder = path.basename(path.dirname(full));
      const raw = fs.readFileSync(full, 'utf8');
      const { data } = matter(raw);
      const sourceContentPath = extractSourceContentPath(data.sourceUrl);
      index[slug] = { filePath: full, localFolder, sourceContentPath };
    }
  };
  walk(docsDir);
  return index;
}

export function splitFrontmatter(md) {
  if (!md.startsWith('---\n')) return { fm: '', body: md };
  const end = md.indexOf('\n---', 4);
  if (end === -1) return { fm: '', body: md };
  const fm = md.slice(0, end + 4);
  const body = md.slice(end + 4).replace(/^\n+/, '');
  return { fm, body };
}

export function toKebab(str) {
  return String(str)
    .normalize('NFKC')
    .toLowerCase()
    .replace(/&/g, ' ')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '');
}

export function readDocFile(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const parsed = matter(content);
  return {
    content,
    body: parsed.content,
    data: parsed.data,
    relativePath: toRelativeDocPath(filePath),
    section: getDocSection(toRelativeDocPath(filePath)),
  };
}

File: scripts/py/src/testim_parity/project.py
"""project レイアウト utilities: docs slug 解決、frontmatter 対応 index 構築。

``scripts/lib/project.mjs`` の port。大半は純粋 path 計算で、一部のみ docs tree を
走査しその結果を ``docs_dir`` ごとにキャッシュする。``matches_section_filter`` は
sidebar 由来の slug 集合を先に引き、失敗時のみ緩いヒューリスティックへ fallback する
(mjs と同じ契約)。
"""

from __future__ import annotations

import logging
import re
import unicodedata
from pathlib import Path
from typing import Any

import frontmatter

_ROOT = Path(__file__).resolve().parents[4]

ROOT_DIR: Path = _ROOT
PROJECT_ROOT: Path = _ROOT
DOCS_DIR: Path = _ROOT / "src" / "content" / "docs"
SIDEBAR_PATH: Path = _ROOT / "docs" / "SIDEBAR_URLS.md"

_log = logging.getLogger(__name__)

_SOURCE_URL_RE = re.compile(
    r"^https://docs\.tricentis\.com/testim/content/([a-z0-9_-]+(?:/[a-z0-9_-]+)*)\.htm$"
)


def _normalize_match_value(value: Any) -> str:
    s = "" if value is None else str(value)
    s = unicodedata.normalize("NFKC", s)
    s = re.sub(r"\s+", " ", s).strip().lower()
    return s


def find_md_files(dir_path: Path | str = DOCS_DIR) -> list[Path]:
    """``dir_path`` 配下の ``.md`` ファイルを再帰的に列挙する。"""
    base = Path(dir_path)
    return sorted(base.rglob("*.md"))


def to_relative_doc_path(file_path: Path | str) -> str:
    """``file_path`` を :data:`ROOT_DIR` 相対の文字列に変換する。"""
    rel = Path(file_path).resolve().relative_to(ROOT_DIR)
    # mjs は OS の path separator を使うので Python 側も同じ挙動にして cross-runtime
    # の conformance 比較が成立するようにしておく。
    return str(rel)


def get_doc_section(relative_path: str) -> str:
    """``src/content/docs/<section>/...`` の section フォルダ名を返す。"""
    parts = relative_path.split("/") if "/" in relative_path else relative_path.split("\\")
    return parts[3] if len(parts) > 3 else ""


def file_path_to_slug(file_path: Path | str, docs_dir: Path | str = DOCS_DIR) -> str:
    """絶対 ``.md`` path を path-based slug に変換する。

    例: ``/.../src/content/docs/overview/testim-overview.md`` →
    ``overview/testim-overview``.
    """
    rel = Path(file_path).resolve().relative_to(Path(docs_dir).resolve())
    return rel.with_suffix("").as_posix()


# モジュールレベル cache (docs_dir 単位)。mjs 挙動と整合させる。test 向けに
# reset API を公開している。plain dict で十分 (key は任意 path、invalidate は
# 明示的)。
# ``_section_cache`` は mjs と同じく **成功時のみ** set を格納する (unknown
# section の ValueError は cache しない — mjs は毎回 sidebar を引き直し warning
# を出すため、conformance 上もその挙動を踏襲する)。キャッシュされた空 set は
# 「sidebar に存在するが slug 0 件の section」という意味で、``if slug_set is not
# None:`` で slug-set 路へ流す。
_section_cache: dict[str, set[str]] = {}
_basename_map_cache: dict[str, dict[str, str | None]] = {}
_slug_index_cache: dict[str, dict[str, dict[str, Any]]] = {}
_resolve_to_full_slug_cache: dict[str, dict[str, str]] = {}


def reset_project_caches_for_test() -> None:
    """モジュールレベル cache を全消去する (tmp ``docs_dir`` を使う test 用)。"""
    _section_cache.clear()
    _basename_map_cache.clear()
    _slug_index_cache.clear()
    _resolve_to_full_slug_cache.clear()


def build_slug_index(docs_dir: Path | str = DOCS_DIR) -> dict[str, dict[str, Any]]:
    """``slug → {categoryFolder, filePath}`` の index を lazy cache 付きで返す。"""
    key = str(Path(docs_dir).resolve())
    cached = _slug_index_cache.get(key)
    if cached is not None:
        return cached
    index: dict[str, dict[str, Any]] = {}
    base = Path(docs_dir)
    for md_file in base.rglob("*.md"):
        slug = file_path_to_slug(md_file, docs_dir)
        category_folder = md_file.parent.name
        index[slug] = {"categoryFolder": category_folder, "filePath": str(md_file)}
    _slug_index_cache[key] = index
    return index


def build_basename_to_path_map(
    docs_dir: Path | str = DOCS_DIR,
) -> dict[str, str | None]:
    """``basename → slug`` の lookup を構築する。重複 basename は ``None``。"""
    key = str(Path(docs_dir).resolve())
    cached = _basename_map_cache.get(key)
    if cached is not None:
        return cached
    slug_index = build_slug_index(docs_dir)
    result: dict[str, str | None] = {}
    for slug in slug_index:
        bn = slug.split("/")[-1]
        if bn in result:
            result[bn] = None  # ambiguous (複数 folder にまたがる basename)
        else:
            result[bn] = slug
    _basename_map_cache[key] = result
    return result


def resolve_to_full_slug(slug: str, docs_dir: Path | str = DOCS_DIR) -> str:
    """省略形 slug を full 形式へ解決。未解決ならそのまま返す。"""
    key = str(Path(docs_dir).resolve())
    cache = _resolve_to_full_slug_cache.setdefault(key, {})
    if slug in cache:
        return cache[slug]
    index = build_slug_index(docs_dir)
    if slug in index:
        resolved = slug
    else:
        basename = slug.split("/")[-1]
        basename_map = build_basename_to_path_map(docs_dir)
        resolved = basename_map.get(basename) or slug
    cache[slug] = resolved
    return resolved


def resolve_slug(input_value: str | None, docs_dir: Path | str = DOCS_DIR) -> str | None:
    """CLI の ``--slug`` 値を full slug へ解決する (path-based 優先)。"""
    if not input_value:
        return None
    index = build_slug_index(docs_dir)
    if input_value in index:
        return input_value
    basename = None if "/" in input_value else input_value
    if not basename:
        return None
    matches = [slug for slug in index if slug.split("/")[-1] == basename]
    if len(matches) == 1:
        _log.warning(
            'Deprecated: basename slug "%s" → "%s". Use the full path-based slug instead.',
            input_value,
            matches[0],
        )
        return matches[0]
    if len(matches) > 1:
        _log.warning(
            'Ambiguous slug "%s" matches multiple paths: %s. Use full path.',
            input_value,
            ", ".join(matches),
        )
    return None


def extract_source_content_path(source_url: str | None) -> str | None:
    """frontmatter ``sourceUrl`` から EN 側の content path を抽出する。

    非文字列・非マッチングは ``None`` を返す。
    """
    if not isinstance(source_url, str):
        return None
    match = _SOURCE_URL_RE.match(source_url)
    if not match:
        return None
    raw = match.group(1)
    return raw[: -len("/index")] if raw.endswith("/index") else raw


def build_docs_index(docs_dir: Path | str = DOCS_DIR) -> dict[str, dict[str, Any]]:
    """EN ``sourceUrl`` 由来の content path も含めた docs index を構築する。

    key は path-based slug (例: ``overview/testim-overview``)。
    """
    index: dict[str, dict[str, Any]] = {}
    base = Path(docs_dir)
    for md_file in base.rglob("*.md"):
        slug = file_path_to_slug(md_file, docs_dir)
        local_folder = md_file.parent.name
        data: dict[str, Any]
        with md_file.open("r", encoding="utf-8") as fh:
            post = frontmatter.load(fh)
            data = dict(post.metadata)
        source_content_path = extract_source_content_path(data.get("sourceUrl"))
        index[slug] = {
            "filePath": str(md_file),
            "localFolder": local_folder,
            "sourceContentPath": source_content_path,
        }
    return index


def split_frontmatter(md: str) -> dict[str, str]:
    """Markdown 文字列を ``fm`` + ``body`` に分割する (mjs と同じ挙動)。

    frontmatter delimiter 無しなら ``{"fm": "", "body": md}`` を返す。
    ``fm`` スライスには末尾の ``\\n---`` を含める (mjs semantics を維持)。
    """
    if not md.startswith("---\n"):
        return {"fm": "", "body": md}
    end = md.find("\n---", 4)
    if end == -1:
        return {"fm": "", "body": md}
    fm = md[: end + 4]
    body = md[end + 4 :].lstrip("\n")
    return {"fm": fm, "body": body}


def to_kebab(value: Any) -> str:
    """mjs ``toKebab`` helper と同じ lossy kebab-case 変換。"""
    s = unicodedata.normalize("NFKC", str(value)).lower()
    s = s.replace("&", " ")
    s = re.sub(r"[^a-z0-9]+", "-", s)
    s = re.sub(r"-+", "-", s)
    s = s.strip("-")
    return s


def read_doc_file(file_path: Path | str) -> dict[str, Any]:
    """単一の .md を読み、``{content, body, data, relativePath, section}`` を返す。"""
    path = Path(file_path)
    content = path.read_text(encoding="utf-8")
    post = frontmatter.loads(content)
    relative_path = to_relative_doc_path(path)
    return {
        "content": content,
        "body": post.content,
        "data": dict(post.metadata),
        "relativePath": relative_path,
        "section": get_doc_section(relative_path),
    }


def matches_section_filter(
    relative_path: str, data: dict[str, Any] | None, section_filter: str | None
) -> bool:
    """``relative_path`` が ``section_filter`` に属するか判定する。

    sidebar 由来の slug 集合を先に consult (成功キャッシュ付き)。失敗時のみ緩い
    ヒューリスティックに fallback する。``sidebar`` は ``madcap_toc`` に依存し、
    ``madcap_toc`` は ``project`` に依存しないので、local import で循環回避する。
    """
    if not section_filter:
        return True

    # Cache は **成功時のみ** 格納する (mjs と同じ semantics)。
    #   - hit (value は空 set 含む) → slug-set 路
    #   - miss (key 未登録) → sidebar を引き、失敗したら warning + heuristic へ
    # mjs は unknown section を cache せず毎回 ``getSectionSlugSet`` を再呼び出し
    # するため、Python も同じく cache しない。Round 2 の sentinel 最適化案は
    # Round 3 codex レビューで divergence として指摘されたため取り消した。
    # (実使用では 1 process あたりの `matchesSectionFilter` 呼び出し回数は高々
    # ページ数 O(10^3) 程度で、unknown section path の重複 disk read は許容範囲。)
    slug_set: set[str] | None = _section_cache.get(section_filter)
    if slug_set is None:
        try:
            from .sidebar import get_section_slug_set

            slug_set = get_section_slug_set(section_filter)
            _section_cache[section_filter] = slug_set
        except ValueError as exc:
            _log.warning("matches_section_filter: %s — falling back to heuristic match", exc)
        except Exception as exc:  # noqa: BLE001 — mjs の catch-all に合わせる
            _log.warning("matches_section_filter: unexpected error: %s", exc)

    # sidebar に存在する section (空 set 含む) は slug-set 路。None は失敗ケースの
    # heuristic fallback。
    if slug_set is not None:
        docs_prefix = "src/content/docs/"
        if relative_path.startswith(docs_prefix):
            rel = relative_path[len(docs_prefix) :].removesuffix(".md")
        else:
            rel = Path(relative_path).stem
        return rel in slug_set

    target = _normalize_match_value(section_filter)
    if not target:
        return True
    candidates = [
        relative_path,
        get_doc_section(relative_path),
        (data or {}).get("category"),
        Path(relative_path).stem,
    ]
    normalized = [_normalize_match_value(c) for c in candidates if c]
    return any(target in candidate for candidate in normalized)


__all__ = [
    "ROOT_DIR",
    "PROJECT_ROOT",
    "DOCS_DIR",
    "SIDEBAR_PATH",
    "find_md_files",
    "to_relative_doc_path",
    "get_doc_section",
    "file_path_to_slug",
    "build_slug_index",
    "build_basename_to_path_map",
    "resolve_to_full_slug",
    "resolve_slug",
    "extract_source_content_path",
    "build_docs_index",
    "split_frontmatter",
    "to_kebab",
    "read_doc_file",
    "matches_section_filter",
    "reset_project_caches_for_test",
]
